### PR TITLE
WIP Fixes(app): #1015 Command Migration: ('DEL', 'EXISTS', 'PERSIST', 'TYPE')

### PR DESCRIPTION
**Pull request by mohit-nagaraj**
### Pull Request Description for Issue #1015

**Description:**
This pull request addresses issue #1015 by migrating the 'DEL', 'EXISTS', 'PERSIST', and 'TYPE' commands to make them compatible across the three protocols supported by DiceDB: RESP, HTTP, and WebSocket. The current implementation of these commands is specific to the RESP protocol. The migration ensures protocol-agnostic behavior by refactoring the evaluation functions to return raw responses instead of protocol-specific responses.

**Changes Made:**
1. **Refactored Evaluation Functions:**
   - Implemented new generic evaluation functions for the mentioned commands in `internal/eval/store_eval.go`.
   - Ensured that the evaluation functions do not include protocol-specific logic.
   - Used raw types without encoding in the return statements.
   - Incorporated errors from `errors/migrated_errors.go` and predefined responses from `internal/clientio/resp.go`.

2. **Command/Worker Specific Changes:**
   - Set the `IsMigrated` flag to true in the command information under `internal/eval/commands.go`.
   - Assigned the new evaluation functions to the `NewEval` parameter in the command structure.
   - Removed the old evaluation functions from `internal/eval/eval.go`.
   - Added migrated commands to the `CommandsMeta` map in `internal/worker/CommandsMeta` with the type `SingleShard`.

3. **Unit and Integration Tests:**
   - Refactored existing unit tests to accommodate the new implementation.
   - Added new unit tests to cover all possible cases.
   - Ran all integration tests to ensure successful migration and correct functionality for all protocols (RESP, HTTP, WebSocket).

**Checklist:**
- [x] Migrated the evalXXX function with the latest definition.
- [x] Updated or added unit tests for the new implementation.
- [x] Ensured all unit tests pass successfully.
- [x] Verified all integration tests pass successfully.

**Additional Notes:**
- Edge cases and protocol-specific optimizations are handled within the wrappers.
- Any questions or concerns about this migration can be mentioned here.

**Related Issues/PRs:**
- Sample implementation for the Get, Set, GetSet, and SetEx commands can be found in a related pull request.

This pull request ensures the command logic is protocol-independent, allowing all three protocols to call the same core functionality seamlessly.

### CODE REVIEW

**Comment by JyotinderSingh** [comment]
Thanks for this PR. It looks like the CI is throwing some errors; please look into it. Additionally, please ensure the following checklist is part of your PR description and all items are checked off. At first glance, I can see that this PR does not add any new integration tests for the resp server.

- [ ] Migrated the evalXXX function with the latest definition
- [ ] Update or add unit tests for the new implementation.
- [ ] All unit tests pass successfully.
- [ ] Ensure all integration tests pass successfully.
- [ ] Move Integration tests for the respective commands under the RESP integration tests directory from Async directory
- [ ] Please validate that the documentation for the respective commands is up to date. If not then consider adding them.

**Comment by mohit-nagaraj** [comment]
Hey @JyotinderSingh I have asked few doubts in the orginal issue could you please help me out with it

**Comment by mohit-nagaraj** [comment]
Hey this pr is ready for review, if anyone can check this out @AshwinKul28 @JyotinderSingh

File: internal/eval/store_eval.go [hunk 1]
```diff
@@ -1932,3 +1932,127 @@ func evalBFINFO(args []string, store *dstore.Store) *EvalResponse {
...
+		}
+	}
+
+	key := args[0]
+	obj := store.Get(key)
+
+	// If the key doesn't exist, return 0
+	if obj == nil {
+		return &EvalResponse{
+			Result: int64(0),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:1997]: Please return clientio.IntegerZero here. That will make it consistent.

File: internal/eval/store_eval.go [hunk 2]
```diff
@@ -1932,3 +1932,127 @@ func evalBFINFO(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  nil,
+		}
+	}
+
+	// If the key has no expiry, return 0
+	_, isExpirySet := dstore.GetExpiry(obj, store)
+	if !isExpirySet {
+		return &EvalResponse{
+			Result: int64(0),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2006]: Same as above, please return clientio.IntegerZero

File: internal/eval/store_eval.go [hunk 3]
```diff
@@ -1932,3 +1932,127 @@ func evalBFINFO(args []string, store *dstore.Store) *EvalResponse {
...
+
+			Error: nil,
+		}
+	}
+
+	// Remove the expiry from the key
+	dstore.DelExpiry(obj, store)
+
+	return &EvalResponse{
+		Result: int64(1),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2016]: Same as above, please return clientio.IntegerOne

**Review by AshwinKul28 — CHANGES_REQUESTED** [review]
Thank you so much @mohit-nagaraj for these efforts. Left few minor comments.

Also we need a documentation updates as well, please have a look at this [sample doc](https://github.com/DiceDB/dice/blob/master/docs/sample_command_docs.md) to make required changes.

**Review by mohit-nagaraj — COMMENTED** [review]
modified & verfied all file changes

**Comment by JyotinderSingh** [comment]
@mohit-nagaraj looks like there are some test failures, could you please take a look?

**Comment by mohit-nagaraj** [comment]
Sure sure looks like some issue due to the recent commit

**Comment by mohit-nagaraj** [comment]
image
Why is the type like this which is sent in response?

**Comment by JyotinderSingh** [comment]
> image Why is the type like this which is sent in response?

@AshwinKul28 are you familiar with this?

**Comment by apoorvyadav1111** [comment]
Hi @mohit-nagaraj , the issue should be fixed in the recent merges, Could you please rebase your changes and let us know for review. We would like to close this soon.
Thanks

**Comment by mohit-nagaraj** [comment]
Sure will check it out tonight

**Comment by mohit-nagaraj** [comment]
Hey @apoorvyadav1111 
Idy why but the issue still persisted on using `clientio.IntegerZero`
for some reason then the types wasnt sent properly still n the unit test was failing for persist function (as the build failed for [bd399a8](https://github.com/mohit-nagaraj/dice/actions/runs/11763985298/job/32768613302#step:7:29))

then i reverted the change which @AshwinKul28 had specified -> [changes](https://github.com/DiceDB/dice/pull/1146#pullrequestreview-2393431607)
then all the tests ran sucessfully without any errors for the commit-> [fc273b7](https://github.com/mohit-nagaraj/dice/actions/runs/11764921604/job/32770636468)

**Comment by mohit-nagaraj** [comment]
@apoorvyadav1111 @AshwinKul28 
does this count as working solution?

**Comment by apoorvyadav1111** [comment]
Hi @mohit-nagaraj , Pleased to know that the issues are resolved, looking forward to close this. Please rebase this to latest master.

Thanks

**Comment by mohit-nagaraj** [comment]
Sure on it

**Comment by apoorvyadav1111** [comment]
Hi @mohit-nagaraj , Please reach out to me on discord if you have any questions.

**Comment by mohit-nagaraj** [comment]
yes yes will do it now, was busy cuz of exams

**Comment by mohit-nagaraj** [comment]
Hey @apoorvyadav1111 theres still a few errors which are coming will ask them over discord

**Comment by apoorvyadav1111** [comment]
Hi @mohit-nagaraj , thanks for the update. You will need to update unit tests with resp values for all the migrated commands to fix the checks. Please let me know or tag me on discord to trigger the runs.

Thanks

**Comment by mohit-nagaraj** [comment]
Ahhh okie so the unit tests have to be updated also then

**Review by apoorvyadav1111 — APPROVED** [review]

**Review by apoorvyadav1111 — APPROVED** [review]
